File: app/src/main/java/com/tkamat/android/youwatch/TopicRecyclerFragment.java
package com.tkamat.android.youwatch;

import android.app.AlertDialog;
import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.constraint.ConstraintLayout;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.support.v4.app.NotificationCompat;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.*;

import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdView;

import java.util.List;
import java.util.UUID;

import static android.content.Context.MODE_PRIVATE;
import static android.content.Context.NOTIFICATION_SERVICE;

public class TopicRecyclerFragment extends Fragment {
    private ConstraintLayout mConstraintLayout;
    private RecyclerView mTopicRecyclerView;
    private TopicAdapter mTopicAdapter;
    private FloatingActionButton mFAB;
    private ProgressBar mProgressBar;
//    private AdView mAdView;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_topic_list, container, false);
        mConstraintLayout = (ConstraintLayout) v.findViewById(R.id.constraint_layout);
        mProgressBar = (ProgressBar) v.findViewById(R.id.progressBar);
        mFAB = (FloatingActionButton) v.findViewById(R.id.add_button);
        mFAB.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (TopicList.get(getActivity()).getEnabledTopics().size() <= 10) {
                    Intent intent = TopicPickerActivity.newIntent(getActivity(), UUID.randomUUID());
                    startActivity(intent);
                } else {
                    Toast.makeText(getActivity(), getString(R.string.limit_reached_toast), Toast.LENGTH_SHORT).show();
                }
            }
        });
//        mAdView = (AdView) v.findViewById(R.id.adView);
//        AdRequest adRequest = new AdRequest.Builder().addTestDevice("BA648A93396B1115DEF0054041E7E8EB").build();
//        mAdView.loadAd(adRequest);
//        mAdView.setAdListener(new AdListener() {
//            @Override
//            public void onAdFailedToLoad(int i) {
//                mAdView.setVisibility(View.GONE);
//                super.onAdFailedToLoad(i);
//            }
//
//            @Override
//            public void onAdLoaded() {
//                mAdView.setVisibility(View.VISIBLE);
//                super.onAdLoaded();
//            }
//        });
        mTopicRecyclerView = (RecyclerView) v.findViewById(R.id.topic_recycler_view);
        mTopicRecyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
        mTopicRecyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                if (dy > 0 || dy < 0 && mFAB.isShown()) {
                    //test
                    mFAB.hide();
                }
            }

            @Override
            public void onScrollStateChanged(RecyclerView recyclerView, int newState) {
                if (newState == RecyclerView.SCROLL_STATE_IDLE) {
                    mFAB.show();
                }

                super.onScrollStateChanged(recyclerView, newState);
            }
        });

        updateUI();
        if (isFirstTime()) {
            final AlertDialog alertDialog = new AlertDialog.Builder(getActivity()).create();
            alertDialog.setTitle(getString(R.string.first_time_title));
            final TextView textView = new TextView(getActivity());
            textView.setText(R.string.first_time_instructions);
            textView.setTextColor(Color.BLACK);
            textView.setTextSize(12);
            alertDialog.setView(textView, 50, 50, 50, 50);
            alertDialog.show();
        }

        Util.scheduleJob(getActivity());
        return v;
    }

    @Override
    public void onResume() {
        updateUI();
        super.onResume();
    }

    public static TopicRecyclerFragment newInstance() {
        TopicRecyclerFragment fragment = new TopicRecyclerFragment();
        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }

    private void updateUI() {
        showViews();

        TopicList topicList = TopicList.get(getActivity());
        List<Topic> topics = topicList.getTopics();

        if (mTopicAdapter == null) {
            mTopicAdapter = new TopicAdapter(topics);
            mTopicRecyclerView.setAdapter(mTopicAdapter);
        } else {
            mTopicAdapter.setmTopics(topics);
            mTopicAdapter.notifyDataSetChanged();
        }
    }

    private void hideViews() {
        mTopicRecyclerView.setVisibility(View.GONE);
        mFAB.setVisibility(View.GONE);
        mProgressBar.setVisibility(View.VISIBLE);
    }

    private void showViews() {
        mTopicRecyclerView.setVisibility(View.VISIBLE);
        mFAB.setVisibility(View.VISIBLE);
        mProgressBar.setVisibility(View.GONE);
    }

    private boolean isFirstTime() {
        SharedPreferences preferences = getActivity().getPreferences(MODE_PRIVATE);
        boolean ranBefore = preferences.getBoolean("RanBefore", false);
        if (!ranBefore) {
            SharedPreferences.Editor editor = preferences.edit();
            editor.putBoolean("RanBefore", true);
            editor.commit();
        }
        return !ranBefore;
    }


    private class TopicHolder extends RecyclerView.ViewHolder {
        private Topic mTopic;
        private TextView mTopicText;
        private TextView mMinimumViews;
        private Switch mSwitch;
        private ConstraintLayout mConstraintLayout;

        public TopicHolder(LayoutInflater inflater, ViewGroup parent) {
            super(inflater.inflate(R.layout.list_item_topic, parent, false));
            mTopicText = (TextView) itemView.findViewById(R.id.topic_text);
            mMinimumViews = (TextView) itemView.findViewById(R.id.minimum_views);
            mSwitch = (Switch) itemView.findViewById(R.id.enabled_switch);
            class SwitchListener implements CompoundButton.OnCheckedChangeListener, View.OnTouchListener {
                boolean userSelect = false;

                @Override
                public boolean onTouch(View view, MotionEvent motionEvent) {
                    userSelect = true;
                    return false;
                }

                @Override
                public void onCheckedChanged(CompoundButton compoundButton, boolean b) {
                    if (userSelect) {
                        if (TopicList.get(getActivity()).getEnabledTopics().size() <= 10) {
                            mTopic.setmEnabled(b);
                            if (b)
                                Toast.makeText(getActivity(), getString(R.string.toast_enabled), Toast.LENGTH_SHORT).show();
                            else
                                Toast.makeText(getActivity(), getString(R.string.toast_disabled), Toast.LENGTH_SHORT).show();
                        } else if (b) {
                            Toast.makeText(getActivity(), getString(R.string.limit_reached_toast), Toast.LENGTH_SHORT).show();
                            userSelect = false;
                            mSwitch.setChecked(false);
                        } else if (!b) {
                            mTopic.setmEnabled(false);
                            Toast.makeText(getActivity(), getString(R.string.toast_disabled), Toast.LENGTH_SHORT).show();
                        }
                    }
                    TopicList.get(getActivity()).updateTopic(mTopic);
                    userSelect = false;
                }
            }
            SwitchListener listener = new SwitchListener();
            mSwitch.setOnCheckedChangeListener(listener);
            mSwitch.setOnTouchListener(listener);

            mConstraintLayout = (ConstraintLayout) itemView.findViewById(R.id.constraint_layout);
            mConstraintLayout.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    hideViews();
                    Intent intent = TopicPickerActivity.newIntent(getActivity(), mTopic.getmID());
                    startActivity(intent);
                }
            });
        }

        public void bind(Topic topic) {
            mTopic = topic;
            mTopicText.setText(mTopic.getmTopicName());
            mMinimumViews.setText(mTopic.getmMinViews() + " views minimum");
            mSwitch.setChecked(mTopic.ismEnabled());
        }
    }

    private class TopicAdapter extends RecyclerView.Adapter<TopicHolder> {
        private List<Topic> mTopics;

        public TopicAdapter(List<Topic> mTopics) {
            this.mTopics = mTopics;
        }

        @Override
        public TopicHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            LayoutInflater layoutInflater = LayoutInflater.from(getActivity());
            return new TopicHolder(layoutInflater, parent);
        }

        @Override
        public void onBindViewHolder(TopicHolder holder, int position) {
            Topic topic = mTopics.get(position);
            holder.bind(topic);
        }

        @Override
        public int getItemCount() {
            return mTopics.size();
        }

        public void setmTopics(List<Topic> mTopics) {
            this.mTopics = mTopics;
        }
    }

}

File: app/src/main/java/com/tkamat/android/youwatch/TopicRecyclerFragment.kt
package com.tkamat.android.youwatch

import android.app.AlertDialog
import android.app.Dialog
import android.content.pm.PackageManager
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.*

import java.util.UUID

import android.content.Context.MODE_PRIVATE
import android.content.SharedPreferences
import android.content.pm.PackageInfo
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.materialswitch.MaterialSwitch
import com.tkamat.android.youwatch.databinding.FragmentTopicListBinding
import com.tkamat.android.youwatch.databinding.ListItemYoutubeTopicBinding

class TopicRecyclerFragment : Fragment() {
    private var _binding: FragmentTopicListBinding? = null
    private val binding get() = _binding!!
    private var topicAdapter: TopicAdapter? = null

    private val isFirstTime: Boolean
        get() {
            var preferences: SharedPreferences? = null
            activity?.let {
                preferences = it.getPreferences(MODE_PRIVATE)
            }
            val ranBefore = preferences?.getBoolean("RanBefore", false)
            if (ranBefore == false) {
                val editor = preferences?.edit()
                editor?.putBoolean("RanBefore", true)
                editor?.commit()
            }
            return !(ranBefore ?: true)
        }

    private val isFirstTimeAfterUpdate: Boolean
        get() {
            if (isFirstTime) {
                return false
            }
            try {
                var packageInfo: PackageInfo? = null
                var preferences: SharedPreferences? = null
                activity?.let {
                    packageInfo = it
                            .packageManager
                            .getPackageInfo(it.packageName, 0)
                    preferences = it.getPreferences(MODE_PRIVATE)
                }
                val currentVersion = packageInfo?.versionCode ?: 0
                val lastRunVersion = preferences?.getInt("LastRunVersion", 7) ?: 0
                return if (currentVersion > lastRunVersion) {
                    val editor = preferences?.edit()
                    editor?.putInt("LastRunVersion", currentVersion)
                    editor?.commit()
                    true
                } else {
                    false
                }
            } catch (e: PackageManager.NameNotFoundException) {
                e.printStackTrace()
            }

            return false
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.fragment_topic_recycler, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.help_button -> showHelpDialog()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentTopicListBinding.inflate(inflater, container, false)

        binding.addButton.setOnClickListener {
            activity?.let {
                if ((TopicList.getInstance(it)?.enabledTopics?.size ?: 0) <= 10) {
                    val intent = TopicPickerActivity.newIntent(it, UUID.randomUUID())
                    startActivity(intent)
                } else {
                    Toast.makeText(it, getString(R.string.limit_reached_toast), Toast.LENGTH_SHORT).show()
                }
            }
        }

        binding.topicRecyclerView.layoutManager = LinearLayoutManager(activity)
        binding.topicRecyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                if (dy > 0) {
                    binding.addButton.hide()
                } else if (dy < 0) {
                    binding.addButton.show()
                }
            }

            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                if (newState == RecyclerView.SCROLL_STATE_IDLE) {
                    binding.addButton.show()
                }
                super.onScrollStateChanged(recyclerView, newState)
            }
        })

        // Setup swipe-to-delete
        setupSwipeToDelete()

        updateUI()

        if (isFirstTime) {
            showHelpDialog()
        }
        if (isFirstTimeAfterUpdate) {
            showUpdateDialog()
        }

        activity?.let {
            Util.scheduleJob(it)
        }
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    private fun setupSwipeToDelete() {
        val swipeHandler = object : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT) {
            private val deleteBackground = ColorDrawable(ContextCompat.getColor(requireContext(), R.color.deleteRed))
            private val deleteIcon = ContextCompat.getDrawable(requireContext(), android.R.drawable.ic_menu_delete)

            override fun onMove(
                recyclerView: RecyclerView,
                viewHolder: RecyclerView.ViewHolder,
                target: RecyclerView.ViewHolder
            ): Boolean = false

            override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
                val position = viewHolder.adapterPosition
                val topics = TopicList.getInstance(requireContext())?.topics
                topics?.getOrNull(position)?.let { topic ->
                    TopicList.getInstance(requireContext())?.deleteTopic(topic.id)
                    updateUI()
                    Toast.makeText(context, "Topic deleted", Toast.LENGTH_SHORT).show()
                }
            }

            override fun onChildDraw(
                c: Canvas,
                recyclerView: RecyclerView,
                viewHolder: RecyclerView.ViewHolder,
                dX: Float,
                dY: Float,
                actionState: Int,
                isCurrentlyActive: Boolean
            ) {
                val itemView = viewHolder.itemView
                val iconMargin = (itemView.height - (deleteIcon?.intrinsicHeight ?: 0)) / 2

                if (dX < 0) {
                    // Swiping left - draw red background
                    deleteBackground.setBounds(
                        itemView.right + dX.toInt(),
                        itemView.top,
                        itemView.right,
                        itemView.bottom
                    )
                    deleteBackground.draw(c)

                    // Draw delete icon
                    deleteIcon?.let {
                        val iconTop = itemView.top + iconMargin
                        val iconBottom = iconTop + it.intrinsicHeight
                        val iconRight = itemView.right - iconMargin
                        val iconLeft = iconRight - it.intrinsicWidth
                        it.setBounds(iconLeft, iconTop, iconRight, iconBottom)
                        it.setTint(Color.WHITE)
                        it.draw(c)
                    }
                }

                super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive)
            }
        }

        ItemTouchHelper(swipeHandler).attachToRecyclerView(binding.topicRecyclerView)
    }

    override fun onResume() {
        updateUI()
        super.onResume()
    }

    private fun updateUI() {
        showViews()

        var topicList: TopicList? = null
        activity?.let {
            topicList = TopicList.getInstance(it)
        }
        val topics = topicList?.topics

        if (topicAdapter == null) {
            topics?.let {
                topicAdapter = TopicAdapter(it)
                binding.topicRecyclerView.adapter = topicAdapter
            }
        } else {
            topics?.let {
                topicAdapter?.setTopics(it)
                topicAdapter?.notifyDataSetChanged()
            }
        }

        activity?.let {
            if (TopicList.getInstance(it)?.topics?.size == 0) {
                binding.emptyScreenHint.visibility = View.VISIBLE
            } else {
                binding.emptyScreenHint.visibility = View.GONE
            }
        }
    }

    private fun hideViews() {
        binding.topicRecyclerView.visibility = View.GONE
        binding.addButton.visibility = View.GONE
    }

    private fun showViews() {
        binding.topicRecyclerView.visibility = View.VISIBLE
        binding.addButton.visibility = View.VISIBLE
    }

    private fun showHelpDialog() {
        val textView = TextView(activity)
        textView.apply {
            setText(R.string.first_time_instructions)
            setTextColor(Color.BLACK)
            textSize = 14f
        }
        val alertDialog = AlertDialog.Builder(activity).create()
        alertDialog.apply {
            setTitle(getString(R.string.first_time_title))
            setView(textView, 48, 48, 48, 12)
            setButton(Dialog.BUTTON_POSITIVE, "OK") { _, _ -> }
            show()
        }
    }

    private fun showUpdateDialog() {
        val textView = TextView(activity)
        textView.apply {
            setText(R.string.update_alert_text)
            setTextColor(Color.BLACK)
            textSize = 14f
        }
        val alertDialog = AlertDialog.Builder(activity).create()
        alertDialog.apply {
            setTitle(getString(R.string.update_alert_title))
            setView(textView, 24, 48, 24, 12)
            setButton(Dialog.BUTTON_POSITIVE, "OK") { _, _ -> }
            show()
        }
    }

    private inner class YoutubeTopicHolder(private val itemBinding: ListItemYoutubeTopicBinding) : RecyclerView.ViewHolder(itemBinding.root) {
        private var topic: YoutubeTopic? = null

        init {
            class SwitchListener : CompoundButton.OnCheckedChangeListener, View.OnTouchListener {
                var userSelect = false

                override fun onTouch(view: View, motionEvent: MotionEvent): Boolean {
                    userSelect = true
                    return false
                }

                override fun onCheckedChanged(compoundButton: CompoundButton, b: Boolean) {
                    activity?.let {
                        if (userSelect) {
                            if (TopicList.getInstance(it)?.enabledTopics?.size ?: 0 <= 10) {
                                topic?.enabled = b
                            } else if (b) {
                                Toast.makeText(it, getString(R.string.limit_reached_toast), Toast.LENGTH_SHORT).show()
                                userSelect = false
                                itemBinding.enabledSwitch.isChecked = false
                            } else if (!b) {
                                topic?.enabled = false
                            }
                        }
                    }
                    activity?.let { a ->
                        topic?.let {t ->
                            TopicList.getInstance(a)?.updateTopic(t)
                        }
                    }
                    userSelect = false
                }
            }

            val listener = SwitchListener()
            itemBinding.enabledSwitch.setOnCheckedChangeListener(listener)
            itemBinding.enabledSwitch.setOnTouchListener(listener)

            itemBinding.constraintLayout.setOnClickListener {
                hideViews()
                activity?.let { a ->
                    topic?.let { t ->
                        val intent = TopicPickerActivity.newIntent(a, t.id)
                        startActivity(intent)
                    }
                }
            }
        }

        fun bind(topic: YoutubeTopic) {
            this.topic = topic
            itemBinding.topicText.text = this.topic?.topicName
            itemBinding.minimumViews.text = "${this.topic?.minViews} views minimum"
            itemBinding.enabledSwitch.isChecked = (this.topic?.enabled ?: true)
        }
    }

    private inner class TopicAdapter(private var topics: List<Topic>?) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
            val itemBinding = ListItemYoutubeTopicBinding.inflate(LayoutInflater.from(parent.context), parent, false)
            return YoutubeTopicHolder(itemBinding)
        }

        override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
            val topic = topics?.get(position)
            val youtubeHolder = holder as? YoutubeTopicHolder
            youtubeHolder?.bind(topic as? YoutubeTopic ?: YoutubeTopic("", 0))
        }

        override fun getItemCount(): Int {
            return topics?.size ?: 0
        }

        fun setTopics(topics: List<Topic>) {
            this.topics = topics
        }
    }

    companion object {

        fun newInstance(): TopicRecyclerFragment {
            val fragment = TopicRecyclerFragment()
            val args = Bundle()
            fragment.arguments = args
            return fragment
        }
    }
}
